keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,936 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This Kep proposes adding a feature (`PodsPerNodeAssignmentPolicy`) that provides
+users with an explicit API and scheduler support for 1:1 pod to node assignment.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Customers trying to deploy batch workloads on Kubernetes have filed multiple
+requests (See [1](https://github.com/kubernetes/kubernetes/issues/68827),
+[2](https://github.com/kubernetes/kubernetes/issues/105287) for details) for 1:1
+pod-to-node assignment. The most common use case for enforcing 1:1 pod to node
+assignment is HPC workloads (MPI and Spark).
+
+Currently there are two approaches to achieve this:
+
+-   Set resource requests to something high enough to prevent scheduling any
+    other pods but daemonset and static pods. This is tricky to do and error
+    prone, especially for smaller nodes. This will not be possible if node size
+    is not known in advance.
+-   Use pod anti affinity. Users will need to define the same set of labels on
+    all pods that needs to be repelled, whether or not they actually request 1:1
+    pod to node assignment, it is also expensive to calculate especially for
+    cluster autoscaler.
+
+A simple and native way to force 1:1 pod-to-node assignment is therefore needed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod
+
+### Non-Goals
+
+-   Find the best fit node for a given pod
+
```

Comment:
Could you please elaborate on the first half? Since we are not adding any preemption logic, I am not sure what should we be documenting here.

For the second half, I added a second bullet point stating that it's a non-goal to avoid those pods from getting preempted.

Reply:
Wei, i am not sure i understand as well, wouldn't it be simpler to just say that this is treated just like any other filter, there is nothing special about it.

Reply:
Oh, I meant if we'd like to **not** favor the pod with `DaemonsetAndStaticPods` policy when it's evaluated as a preepmption candidate (suppose there are multiple preemption options). For example, to schedule a high-priority pod-X, possible preemption options are:

1. preempt a regular pod-1 on node-A
2. preempt a pod-2 with `DaemonsetAndStaticPods` policy on node-B

My point is that, in current defaultpreemption logic, it's random to pick pod-1 or pod-2 as the final preemption, but maybe we can prefer to protect pod-2 more  as it may not have a good chance to be scheduled next time (due to its coexist policy). Generally, make a pod with coexist policy less disruptive upon premmption can make the feature more useful.

We may or may not pursue it, just write it down to manage the user's expectation.

Reply:
Or if we just skip the preemption of `pod-per-node` named pod, since the landed node is very likely be picked as the candidate node(according to `pickOneNodeForPreemption` rule, like 
```
  // 1. A node with minimum number of PDB violations.
  // 2. A node with minimum highest priority victim is picked.
  // 3. Ties are broken by sum of priorities of all victims.
  // 4. If there are still ties, node with the minimum number of victims is picked.
  // 5. If there are still ties, node with the latest start time of all highest priority victims is picked.
  // 6. If there are still ties, the first such node is picked (sort of randomly).
  ```
), which may leading to the victim pod unschedulable. 

Unless the scheduling pod is also a `pod-per-node` pod.

Reply:
ok, we could modify the priority to not prefer pods with `DaemonsetAndStaticPods` between steps 3 and 4

Reply:
Shouldn't this go before 3? My understanding is that sum of priorities for nodes having a pod with `DaemonsetAndStaticPods` might also be affected by the low number of pods on such nodes. Is this correct?

Also, could it be better to keep the default behavior as is, and to avoid such pods from getting preempted, users should give them a high enough priority to avoid them from getting preempted similar to a regular pod? I am afraid this might end up with lots of special cases hard coded.

Reply:
> Shouldn't this go before 3? My understanding is that sum of priorities for nodes having a pod with DaemonsetAndStaticPods might also be affected by the low number of pods on such nodes. Is this correct?

Hard to tell, it depends on the range of priorities created.

> Also, could it be better to keep the default behavior as is, and to avoid such pods from getting preempted, users should give them a high enough priority to avoid them from getting preempted similar to a regular pod? I am afraid this might end up with lots of special cases hard coded.

This is just a preference, if they setup the workloads such that those pods have higher priority, then they wouldn't even be considered for preemption at all.

In general, I don't have strong opinion either way, but I wouldn't mind having an explicit preference to avoid preempting those pods first in case they have lower priority than the incoming pod.

Reply:
As I mentioned earlier, I'm just throwing out this potential idea. I'm fine with keep the preemption logic unchanged in alpha (document the behavior is always needed), and see if we can get feedback or real workload data to help us make the current decision on beta.

Reply:
ok, then just mention that this will not impact preferences on selecting candidate nodes upon preemption.